### PR TITLE
Reporting update

### PR DESCRIPTION
**Pull request by camagallen**
# Pull Request Submission

## About
Update to combine reports and stratified_reports, as well as to 

Resolves: [AB#14752](https://dev.azure.com/kensci/91234890-e1c9-4679-971f-09afd8516db1/_workitems/edit/14752), [AB#13810](https://dev.azure.com/kensci/91234890-e1c9-4679-971f-09afd8516db1/_workitems/edit/13810)

## Integration Issues
NA

## How was this update tested?
Through pytest (new test file added) and through actively re-running the tutorial notebooks.

## Checklist
- [X] I have read the [Contributing](CONTRIBUTING.md) guidelines.
- [ ] I performed a self-review of my own code.
- [X] The CHANGELOG and version.json have been updated.
- [X] Documentation has been updated to reflect the changes.
    - [X] READMEs, Tutorials/Examples, and links have been updated as appropriate.
- [X] Tests have been added to prove that the fix is effective or that the feature works.
- [ ] Pytest was run; new and existing unit tests pass locally with my changes.
- [X] I have made note of any breakages or dependencies (if applicable).
- [X] Any dependent changes have been merged and published in downstream modules.
- [X] These changes generate no new warnings.

### Updates required to combine reports.py and stratified_reports.py into one file
* README.md 
* fairmlhealth/__report_processing.py
* fairmlhealth/model_comparison.py
* fairmlhealth/utils.py (preprocessing is now contained in __report_processing.py)
* templates/*
* tutorials_and_examples/*
* test/test_reports.py

### Refactor/cleanup updates that probably should have waited for a later PR
* fairmlhealth/__classification_metrics.py
* docs/development/PULL_REQUEST_TEMPLATE.md
* test/development/README.md

### CODE REVIEW

File: fairmlhealth/__classification_metrics.py [hunk 1]
```diff
@@ -82,30 +81,41 @@ def balanced_accuracy(y_true, y_pred):
...
+        Related metrics such as MAPE (Mean Absolute Percentage Error) or SMAPE
+        may be invalid for asymmetrical prediction ranges with negative values.
+    """
+    if y_range is not None:
+        if not isinstance(y_range, (int, float)):
+            err = "Invalid y_range. Must be int or float describing true range."
+            raise ValueError(err)
+    else:
+        y_range = max(epsilon(), np.max(y_true) - np.min(y_true))
+    rmae = np.mean(np.abs(y_true - y_pred))/y_range
```
Comment (kenviku) [inline fairmlhealth/__classification_metrics.py:112]: nit: `scmae`?

File: fairmlhealth/reports.py [hunk 2]
```diff
@@ -4,295 +4,148 @@
...
-    return dt_vals
+    if target_labels is None:
+        target_labels = [f"target = {t}" for t in set(y_true)]
+    report = classification_report(y_true, y_pred, output_dict=True,
+                                             target_names=target_labels)
+    report = pd.DataFrame(report).transpose()
+    # Move accuracy to separate row
+    accuracy = report.loc['accuracy', :]
+    report.drop('accuracy', inplace=True)
+    report.loc['accuracy', 'accuracy'] = accuracy[0]
```
Comment (kenviku) [inline fairmlhealth/reports.py:87]: moving it to last row?

File: fairmlhealth/reports.py [hunk 3]
```diff
@@ -362,16 +214,671 @@ def color_ratios(row):
...
+    df = stratified_preprocess(X, y_true, y_pred, y_prob, features=features)
+    yt, yh, yp = y_cols(df)['col_names'].values()
+    pred_cols = [n for n in [yt, yh, yp] if n is not None]
+    stratified_features = [f for f in df.columns.tolist() if f not in pred_cols]
+    if any(y is None for y in [yt, yh]):
+        raise ValidationError("Cannot generate report with undefined targets")
+    #
+    res = []
+    errs = {}
+    for f in stratified_features:
```
Comment (kenviku) [inline fairmlhealth/reports.py:456]: may be separate these sections into a function on its own? I would assume that lot of other cosmetic changes would be required in future for the feature-specific attributes.

**Review by kenviku — COMMENTED** [review]
Added some comments.

File: fairmlhealth/reports.py (hunk 2 above)
Comment (camagallen) [reply fairmlhealth/reports.py:87]: yes

**Review by kenviku — APPROVED** [review]
Few minor comments. LGTM

File: fairmlhealth/reports.py (hunk 3 above)
Comment (camagallen) [reply fairmlhealth/reports.py:411]: Added as issue.

File: fairmlhealth/__classification_metrics.py [hunk 4]
```diff
@@ -51,7 +51,7 @@ def check_result(res, metric_name):
 
 def epsilon():
     """ error value used to prevent 'division by 0' errors """
-    return 0.0000000001
+    return np.finfo(np.float64).eps
```
[the remark this answers is not in the document]
Comment (camagallen) [reply fairmlhealth/__classification_metrics.py:43]: Simply choosing a standardized value.

**Pull request merged**